mne/preprocessing/hfc.py

```diff
@@ -0,0 +1,128 @@
+# Authors: George O'Neill <g.o'neill@ucl.ac.uk>
+#
+# License: BSD-3-Clause
+
+import numpy as np
+from .maxwell import _prep_mf_coils, _sss_basis
+from ..io.pick import _picks_to_idx
+from ..io.proj import Projection
+from ..utils import verbose
+
+
+@verbose
+def compute_proj_hfc(info, order=1, picks='meg', exclude='bads',
+                     ignore_ref=True, accuracy='accurate', verbose=None):
+    """Generate projectors to perform homogeneous/harmonic correction to data.
+
+    Remove evironmental fields from magentometer data by assuming it is
+    explained as a homogeneous :footcite:`TierneyEtAl2021` or harmonic field
+    :footcite:`TierneyEtAl2022`. Useful for arrays of OPMs.
+
+    Parameters
+    ----------
+    info : instance of Info
+        Info from an instance of Raw (preferable).
+    order : int
+        The order of the spherical harmonic basis set to use. Set to 1 to use
+        only the homogeneous field component (default), 2 to add gradients, 3
+        to add quadrature terms etc.
+    picks : str | array_like | slice | None
+        Channels to include. Default of ``'meg'`` will select all non-reference
+        MEC channels.
+    exclude : list | 'bads'
+        List of channels to exclude from HFC, only used when picking
+        based on types (e.g., exclude="bads" when picks="meg").
+        Specify ``'bads'`` (the default) to exclude all channels marked as bad.
+    ignore_ref : bool
+        Specify whether reference MEG channels should be ignored when
+        calculating the basis set (default: ``True``).
+    accuracy : str
+        Can be ``"point"``, ``"normal"`` or ``"accurate"`` (default), defines
+        which level of coil definition accuracy is used to generate model.
+    %(verbose)s
+
+    Returns
+    -------
+    projs : list of Projection
+        List of projection vectors.
+
+    Notes
+    -----
+    To apply the projectors to a dataset, use ``inst.add_proj(projs).apply
+    _proj()``.
+
+    References
+    ----------
+    .. footbibliography::
+    """
+    idx = _picks_to_idx(info, picks, exclude=exclude)
+    if not ignore_ref:
+        idx = np.union1d(idx,
+                         _picks_to_idx(info, picks='ref_meg', exclude=exclude))
+    basis, channels = _generate_basis_set(info, idx, order,
+                                          accuracy, ignore_ref)
```

Comment:
Something still seems off about this. You shouldn't need to do anything special after `_picks_to_idx`, this should just work to get you what you want:
```suggestion
    picks = _picks_to_idx(info, picks, none='meg', with_ref_meg=not ignore_ref,
                          exclude=exclude)
    basis, channels = _generate_basis_set(info, picks, order, accuracy)
```
i.e., let `picks_to_idx` take care of picking the set of channels that should be used, then use all of them thereafter (no need to pass `ignore_ref` anymore).

At least this is the case in principle. The following code seems to suggest that *we have a bug with _picks_to_idx where with_ref_meg=False does not work properly* but I think we should fix that separately:
```
>>> raw = mne.io.read_raw_ctf(mne.datasets.testing.data_path() / 'CTF' / 'catch-alp-good-f.ds')
>>> raw.info
<Info | 16 non-empty values
 bads: []
 ch_names: STIM, TIME01, BG1-606, BG2-606, BG3-606, BP1-606, BP2-606, ...
 chs: 1 Stimulus, 3 misc, 29 Reference Magnetometers, 151 Magnetometers, 60 EEG
...
>>> len(mne.io.pick._picks_to_idx(raw.info, 'mag', with_ref_meg=False))
151
>>> len(mne.io.pick._picks_to_idx(raw.info, 'meg', with_ref_meg=False))
180
>>> len(mne.io.pick._picks_to_idx(raw.info, 'ref_meg', with_ref_meg=True))
29
```
If agree I can open a quick PR to fix this `_picks_to_idx` behavior (hopefully it doesn't break too much stuff!) and then you can use it here.